### PR TITLE
[Merged by Bors] - feat: port Data.Finsupp.Basic

### PR DESCRIPTION
**Pull request by j-loreaux**
porting notes:
1. `simps` was not working in various places due to the `noncomputable`, so I had to manually add some lemmas it would have generated. See here: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/finding.20.60simps.60.20lemmas
2. Lean 4 really had trouble inferring arguments to `mapRange_comp` for some reason as compared to Lean 3
3. In several places I had to manually supply decidability instances (from `Classical`). This should probably be investigated.
4. It looks like the `Sigma.mk.inj` auto-generated lemma was never created. Why?
5. In one place, `ext` didn't fire the same way in Lean 3 versus Lean 4; I just copied the output of `ext?` from Lean 3, but we should understand what went wrong.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Finsupp/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,1924 @@
...
+
+theorem sum_option_index_smul [Semiring R] [AddCommMonoid M] [Module R M] (f : Option α →₀ R)
+    (b : Option α → M) :
+    (f.sum fun o r => r • b o) = f none • b none + f.some.sum fun a r => r • b (Option.some a) :=
+  f.sum_option_index _ (fun _ => zero_smul _ _) fun _ _ _ => add_smul _ _ _
+#align finsupp.sum_option_index_smul Finsupp.sum_option_index_smul
+
+end Option
+
+/-! ### Declarations about `filter` -/
```
Comment (jcommelin) [inline Mathlib/Data/Finsupp/Basic.lean:896]:
```suggestion
/-! ### Declarations about `Finsupp.filter` -/
```

File: Mathlib/Data/Finsupp/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,1924 @@
...
+    classical
+      rw [single_apply] at ht2⊢
+      split_ifs  at ht2⊢
+      · exact Finset.mem_singleton_self _
+      · exact (t ht2.symm).elim
+#align finsupp.frange_single Finsupp.frange_single
+
+end Frange
+
+/-! ### Declarations about `subtype_domain` -/
```
Comment (jcommelin) [inline Mathlib/Data/Finsupp/Basic.lean:1039]:
```suggestion
/-! ### Declarations about `Finsupp.subtypeDomain` -/
```

File: Mathlib/Data/Finsupp/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,1924 @@
...
+#align finsupp.sum_curry_index Finsupp.sum_curry_index
+
+/-- Given a finitely supported function `f` from `α` to the type of
+finitely supported functions from `β` to `M`,
+`uncurry f` is the "uncurried" finitely supported function from `α × β` to `M`. -/
+protected def uncurry (f : α →₀ β →₀ M) : α × β →₀ M :=
+  f.sum fun a g => g.sum fun b c => single (a, b) c
+#align finsupp.uncurry Finsupp.uncurry
+
+/-- `Finsupp_prod_equiv` defines the `Equiv` between `((α × β) →₀ M)` and `(α →₀ (β →₀ M))` given by
```
Comment (jcommelin) [inline Mathlib/Data/Finsupp/Basic.lean:1279]:
```suggestion
/-- `finsuppProdEquiv` defines the `Equiv` between `((α × β) →₀ M)` and `(α →₀ (β →₀ M))` given by
```

**Review by jcommelin — COMMENTED** [review]
Only some comments about comments (-;

**Comment by j-loreaux** [comment]
FYI: I rebased in order to incorporate #1958 from master.

File: Mathlib/Data/Finsupp/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,1924 @@
...
+    left_inv := fun x =>
+      by
+      rw [← mapRange_comp _ _ _ _] <;> simp_rw [AddEquiv.symm_comp_self]
+      · exact mapRange_id _
+      · rfl
+    right_inv := fun x =>
+      by
+      rw [← mapRange_comp _ _ _ _] <;> simp_rw [AddEquiv.self_comp_symm]
+      · exact mapRange_id _
+      · rfl }
```
Comment (jcommelin) [inline Mathlib/Data/Finsupp/Basic.lean:275]:
```suggestion
def mapRange.addEquiv (f : M ≃+ N) : (α →₀ M) ≃+ (α →₀ N) := where
  mapRange.addMonoidHom f.toAddMonoidHom with
  toFun := (mapRange f f.map_zero : (α →₀ M) → α →₀ N)
  invFun := (mapRange f.symm f.symm.map_zero : (α →₀ N) → α →₀ M)
  left_inv := fun x => by
    rw [← mapRange_comp _ _ _ _] <;> simp_rw [AddEquiv.symm_comp_self]
    · exact mapRange_id _
    · rfl
  right_inv := fun x => by
    rw [← mapRange_comp _ _ _ _] <;> simp_rw [AddEquiv.self_comp_symm]
    · exact mapRange_id _
    · rfl
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**